docs/computing/containers/overview.md

```diff
@@ -1,71 +1,260 @@
+```
+
+We can also clean the cache directory if necessary:
+
+```bash
+apptainer cache clean
+```
+
+### Virtual memory limit
+
+The virtual memory is limit on Puhti and Mahti login nodes is quite small (10 GiB) and should be increased to the hard limit (24 GiB).
+Exceeding the virtual memory limit causes memory errors during build.
+You can query the current virtual memory limit using `ulimit -v` and the hard limit using `ulimit -Hv`.
+We can set the virtual memory limit to the hard limit as follows:
+
+```bash
+ulimit -v $(ulimit -Hv)
+```
+
+If your build runs out of virtual memory during the build on the login node, you should use an interactive job where virtual memory is limited to the amount of memory reserved for the job.
+
+### Bind mounting temporary directory
+
+By default Apptainer bind mounts the host's `/tmp` to `/tmp` in the build environment.
+However, the size of `/tmp` is limited on Puhti and Mahti, thus, we bind mount the local disk (`$TMPDIR`) to `/tmp` to avoid running out of disk space as follows: `--bind="$TMPDIR:/tmp"`.
+
+### Building SIF image from existing Docker or OCI image
+
+We can obtain existing container images from a container registry by pulling them.
+Apptainer will convert them from Docker or OCI format into the Singularity Image Format (SIF).
+
+```bash
+apptainer build rockylinux.sif docker://docker.io/rockylinux/rockylinux:8.10
+```
+
+### Building SIF image from definition file
+
+Apptainer definition files are written using the `.def` file extension.
+Here is a simple example of container definition:
+
+```sh title="container.def"
+Bootstrap: docker
+From: docker.io/rockylinux/rockylinux:8.10
+
+%post
+    # Replace the failing commands with always succeeding dummies.
+    cp /usr/bin/true /usr/sbin/useradd
+    cp /usr/bin/true /usr/sbin/groupadd
+
+    # Continue to install software into the container normally.
+    dnf -y update  # would fail without the dummies
+```
+
+We can invoke Apptainer to build the container (`container.sif`) from the definition file (`container.def`) using fakeroot as follows:
+
+```bash
+apptainer build --fakeroot --bind="$TMPDIR:/tmp" container.sif container.def
+```
+
+
+### Developing with interactive sandbox
+
+We can also build Apptainer sandboxes with fakeroot.
+Sandboxes are useful for interactive development of containers.
+The sandbox must be created on the local disk (`$TMPDIR`), not on the Lustre parallel file system.
+
+We can initialize a sandbox from a base image as follows:
+
+```bash
+apptainer build --fakeroot --sandbox "$TMPDIR/rockylinux" docker://docker.io/rockylinux/rockylinux:8.10
+```
+
+Then we can run a shell in the sandbox to install software into it:
+
+```bash
+apptainer shell --fakeroot --writable --contain --cleanenv --bind="$TMPDIR:/tmp" "$TMPDIR/rockylinux"
+```
+
+We can use the same tricks to replace the failing commands in the sandbox:
+
+```bash
+cp /usr/bin/true /usr/sbin/useradd
+cp /usr/bin/true /usr/sbin/groupadd
+```
```

Comment:
Suggested continuation from here:

After this we can run any installation commands needed, such as `apt install` or `pip install` interactively in the sandbox shell. Once you are done, exit the shell and convert the sandbox into a `.sif` file:

```bash
apptainer build --fakeroot --bind="$TMPDIR:/tmp" container.sif "$TMPDIR/rockylinux"
```

Jaan: please double check that this is the right command, I just made it up from my head ;-)